trie/database.go

```diff
@@ -106,23 +73,19 @@ func NewDatabase(diskdb ethdb.Database, config *Config) *Database {
 		log.Crit("Both 'hash' and 'path' mode are configured")
 	}
 	if config.PathDB != nil {
-		db.backend = pathdb.New(diskdb, config.PathDB)
+		db.backend = config.PathDB.New(diskdb)
+	} else if config.HashDB != nil {
+		db.backend = config.HashDB.New(diskdb, mptResolver{})
 	} else {
-		db.backend = hashdb.New(diskdb, config.HashDB, mptResolver{})
+		db.backend = HashDefaults.HashDB.New(diskdb, mptResolver{})
 	}
 	return db
 }
 
 // Reader returns a reader for accessing all trie nodes with provided state root.
 // An error will be returned if the requested state is not available.
-func (db *Database) Reader(blockRoot common.Hash) (Reader, error) {
-	switch b := db.backend.(type) {
-	case *hashdb.Database:
-		return b.Reader(blockRoot)
-	case *pathdb.Database:
-		return b.Reader(blockRoot)
-	}
-	return nil, errors.New("unknown backend")
```

Comment:
it is not needed to fork on the concrete type of the db backend since the interface supports it. I also applied this upstream.
There has also been some refactoring upstream here, which I partially applied to make things a bit more simple (provides a place for the interfaces)